File: android/app/src/main/java/com/boxboxnow/app/imu/ImuCalibrator.kt
package com.boxboxnow.app.imu

import com.boxboxnow.app.models.GPSSample
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlin.math.sqrt

/**
 * IMU calibration matching iOS ImuCalibrator / web calibration.ts.
 *
 * Phase 1 (sampling) — kart stationary: capture 100 gravity samples to remove bias.
 * Phase 2 (ready)    — kart moving >15 km/h: correlate heading+accel to align axes.
 */
class ImuCalibrator {
    enum class Phase { IDLE, SAMPLING, READY, ALIGNED }

    private data class Vec3(var x: Double, var y: Double, var z: Double) {
        companion object {
            val ZERO get() = Vec3(0.0, 0.0, 0.0)
        }
    }

    private val _phase = MutableStateFlow(Phase.IDLE)
    val phase = _phase.asStateFlow()

    private val _progress = MutableStateFlow(0.0)
    val progress = _progress.asStateFlow()

    private val gravSamples = mutableListOf<Vec3>()
    private val headingSamples = mutableListOf<Pair<Double, Double>>()
    private var gravity = Vec3.ZERO
    private var rotationMatrix: DoubleArray? = null

    private val staticSampleCount = 100
    private val alignSpeedKmh = 15.0
    private val alignSampleCount = 10
    private val minAccelMagnitude = 0.05

    fun startCalibration() {
        gravSamples.clear()
        headingSamples.clear()
        gravity = Vec3.ZERO
        rotationMatrix = null
        _progress.value = 0.0
        _phase.value = Phase.SAMPLING
    }

    fun skipAlignment() {
        if (_phase.value != Phase.READY) return
        rotationMatrix = doubleArrayOf(1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0)
        _phase.value = Phase.ALIGNED
    }

    fun reset() {
        _phase.value = Phase.IDLE
        _progress.value = 0.0
        gravSamples.clear()
        headingSamples.clear()
        gravity = Vec3.ZERO
        rotationMatrix = null
    }

    /** Processes a sample and returns it with gravity removed + axes rotated. */
    fun calibrate(sample: GPSSample): GPSSample {
        when (_phase.value) {
            Phase.IDLE -> return sample

            Phase.SAMPLING -> {
                gravSamples.add(Vec3(sample.gForceX, sample.gForceY, sample.gForceZ))
                _progress.value = gravSamples.size.toDouble() / staticSampleCount
                if (gravSamples.size >= staticSampleCount) {
                    computeGravity()
                    _progress.value = 1.0
                    _phase.value = Phase.READY
                }
                return sample
            }

            Phase.READY -> {
                val s = sample.copy(
                    gForceX = sample.gForceX - gravity.x,
                    gForceY = sample.gForceY - gravity.y,
                    gForceZ = sample.gForceZ - gravity.z,
                )
                if (s.speedKmh > alignSpeedKmh) {
                    val mag = sqrt(s.gForceX * s.gForceX + s.gForceY * s.gForceY)
                    if (mag >= minAccelMagnitude) {
                        headingSamples.add(s.gForceX to s.gForceY)
                        if (headingSamples.size >= alignSampleCount) {
                            buildRotationMatrix()
                            _phase.value = Phase.ALIGNED
                        }
                    }
                }
                return s
            }

            Phase.ALIGNED -> {
                val noGrav = Vec3(
                    sample.gForceX - gravity.x,
                    sample.gForceY - gravity.y,
                    sample.gForceZ - gravity.z,
                )
                val r = applyRotation(noGrav)
                return sample.copy(gForceX = r.x, gForceY = r.y, gForceZ = r.z)
            }
        }
    }

    private fun computeGravity() {
        val n = gravSamples.size.toDouble()
        gravity = Vec3(
            gravSamples.sumOf { it.x } / n,
            gravSamples.sumOf { it.y } / n,
            gravSamples.sumOf { it.z } / n,
        )
    }

    private fun buildRotationMatrix() {
        val gLen = sqrt(gravity.x * gravity.x + gravity.y * gravity.y + gravity.z * gravity.z)
        if (gLen < 0.001) {
            rotationMatrix = doubleArrayOf(1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0)
            return
        }
        val up = Vec3(-gravity.x / gLen, -gravity.y / gLen, -gravity.z / gLen)

        var fx = 0.0
        var fy = 0.0
        for ((gx, gy) in headingSamples) { fx += gx; fy += gy }
        val fLen2d = sqrt(fx * fx + fy * fy)
        if (fLen2d < 0.001) {
            rotationMatrix = doubleArrayOf(1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0)
            return
        }

        var fwd = Vec3(fx / fLen2d, fy / fLen2d, 0.0)
        val dot = fwd.x * up.x + fwd.y * up.y + fwd.z * up.z
        fwd = Vec3(fwd.x - dot * up.x, fwd.y - dot * up.y, fwd.z - dot * up.z)
        val fwdLen = sqrt(fwd.x * fwd.x + fwd.y * fwd.y + fwd.z * fwd.z)
        if (fwdLen < 0.001) {
            rotationMatrix = doubleArrayOf(1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0)
            return
        }
        fwd = Vec3(fwd.x / fwdLen, fwd.y / fwdLen, fwd.z / fwdLen)

        val right = Vec3(
            fwd.y * up.z - fwd.z * up.y,
            fwd.z * up.x - fwd.x * up.z,
            fwd.x * up.y - fwd.y * up.x,
        )

        // Column-major
        rotationMatrix = doubleArrayOf(
            right.x, fwd.x, up.x,
            right.y, fwd.y, up.y,
            right.z, fwd.z, up.z,
        )
    }

    private fun applyRotation(v: Vec3): Vec3 {
        val r = rotationMatrix ?: return v
        return Vec3(
            r[0] * v.x + r[3] * v.y + r[6] * v.z,
            r[1] * v.x + r[4] * v.y + r[7] * v.z,
            r[2] * v.x + r[5] * v.y + r[8] * v.z,
        )
    }
}

File: BoxBoxNow/BoxBoxNow/Services/ImuCalibrator.swift
import Foundation
import Combine

/// IMU calibration matching the web's calibration.ts logic.
///
/// Phase 1 — Static (kart stationary): captures 100 gravity samples to remove bias.
/// Phase 2 — Dynamic (kart moving >15 km/h): uses GPS heading + acceleration
///           correlation to align device axes to vehicle axes (lateral/longitudinal/vertical).
final class ImuCalibrator: ObservableObject {
    enum Phase: String, Equatable {
        case idle, sampling, ready, aligned
    }

    struct Vec3 {
        var x, y, z: Double
        static let zero = Vec3(x: 0, y: 0, z: 0)
    }

    @Published var phase: Phase = .idle
    @Published var progress: Double = 0  // 0-1 during sampling

    private var gravSamples: [Vec3] = []
    private var headingSamples: [(gx: Double, gy: Double)] = []
    private var gravity: Vec3 = .zero
    private var rotationMatrix: [Double]? = nil  // 3x3 column-major

    private let staticSampleCount = 100
    private let alignSpeedKmh = 15.0
    private let alignSampleCount = 10
    private let minAccelMagnitude = 0.05  // G — minimum to detect braking/accel

    // MARK: - Public API

    func startCalibration() {
        gravSamples.removeAll()
        headingSamples.removeAll()
        gravity = .zero
        rotationMatrix = nil
        progress = 0
        phase = .sampling
    }

    func skipAlignment() {
        guard phase == .ready else { return }
        rotationMatrix = [1, 0, 0, 0, 1, 0, 0, 0, 1]  // identity
        phase = .aligned
    }

    func reset() {
        phase = .idle
        progress = 0
        gravSamples.removeAll()
        headingSamples.removeAll()
        gravity = .zero
        rotationMatrix = nil
    }

    /// Process a sample through the calibration pipeline.
    /// Returns the sample with calibrated g-force values.
    func calibrate(sample: GPSSample) -> GPSSample {
        var s = sample

        switch phase {
        case .idle:
            return s

        case .sampling:
            // Collect static gravity samples
            gravSamples.append(Vec3(x: s.gForceX, y: s.gForceY, z: s.gForceZ))
            progress = Double(gravSamples.count) / Double(staticSampleCount)

            if gravSamples.count >= staticSampleCount {
                computeGravity()
                progress = 1.0
                phase = .ready
            }
            return s

        case .ready:
            // Remove gravity
            s.gForceX -= gravity.x
            s.gForceY -= gravity.y
            s.gForceZ -= gravity.z

            // Collect heading samples when moving fast enough
            if s.speedKmh > alignSpeedKmh {
                let mag = sqrt(s.gForceX * s.gForceX + s.gForceY * s.gForceY)
                if mag >= minAccelMagnitude {
                    headingSamples.append((gx: s.gForceX, gy: s.gForceY))

                    if headingSamples.count >= alignSampleCount {
                        buildRotationMatrix()
                        phase = .aligned
                    }
                }
            }
            return s

        case .aligned:
            // Full transform: remove gravity + rotate to vehicle axes
            let noGrav = Vec3(
                x: s.gForceX - gravity.x,
                y: s.gForceY - gravity.y,
                z: s.gForceZ - gravity.z
            )
            let rotated = applyRotation(noGrav)
            s.gForceX = rotated.x
            s.gForceY = rotated.y
            s.gForceZ = rotated.z
            return s
        }
    }

    // MARK: - Private

    private func computeGravity() {
        let n = Double(gravSamples.count)
        gravity = Vec3(
            x: gravSamples.reduce(0) { $0 + $1.x } / n,
            y: gravSamples.reduce(0) { $0 + $1.y } / n,
            z: gravSamples.reduce(0) { $0 + $1.z } / n
        )
    }

    private func buildRotationMatrix() {
        let gLen = sqrt(gravity.x * gravity.x + gravity.y * gravity.y + gravity.z * gravity.z)
        guard gLen > 0.001 else {
            rotationMatrix = [1, 0, 0, 0, 1, 0, 0, 0, 1]
            return
        }

        // "Up" axis = normalize(-gravity)
        let up = Vec3(x: -gravity.x / gLen, y: -gravity.y / gLen, z: -gravity.z / gLen)

        // "Forward" axis: average acceleration direction during braking/accel
        var fx = 0.0, fy = 0.0
        for s in headingSamples {
            fx += s.gx
            fy += s.gy
        }
        let fLen2d = sqrt(fx * fx + fy * fy)
        guard fLen2d > 0.001 else {
            rotationMatrix = [1, 0, 0, 0, 1, 0, 0, 0, 1]
            return
        }

        // Forward in device coords (projected)
        var fwd = Vec3(x: fx / fLen2d, y: fy / fLen2d, z: 0)

        // Orthogonalize forward w.r.t. up: fwd = fwd - (fwd·up)*up
        let dot = fwd.x * up.x + fwd.y * up.y + fwd.z * up.z
        fwd = Vec3(x: fwd.x - dot * up.x, y: fwd.y - dot * up.y, z: fwd.z - dot * up.z)
        let fwdLen = sqrt(fwd.x * fwd.x + fwd.y * fwd.y + fwd.z * fwd.z)
        guard fwdLen > 0.001 else {
            rotationMatrix = [1, 0, 0, 0, 1, 0, 0, 0, 1]
            return
        }
        fwd = Vec3(x: fwd.x / fwdLen, y: fwd.y / fwdLen, z: fwd.z / fwdLen)

        // "Right" (lateral) = forward × up
        let right = Vec3(
            x: fwd.y * up.z - fwd.z * up.y,
            y: fwd.z * up.x - fwd.x * up.z,
            z: fwd.x * up.y - fwd.y * up.x
        )

        // Column-major: maps device coords → vehicle coords
        // x = lateral (right+), y = longitudinal (forward+), z = vertical (up+)
        rotationMatrix = [
            right.x, fwd.x, up.x,
            right.y, fwd.y, up.y,
            right.z, fwd.z, up.z,
        ]
    }

    private func applyRotation(_ v: Vec3) -> Vec3 {
        guard let R = rotationMatrix else { return v }
        // Column-major: R[col*3+row]
        return Vec3(
            x: R[0] * v.x + R[3] * v.y + R[6] * v.z,
            y: R[1] * v.x + R[4] * v.y + R[7] * v.z,
            z: R[2] * v.x + R[5] * v.y + R[8] * v.z
        )
    }
}
